format/bookmark/bookmark.md

```diff
@@ -0,0 +1,42 @@
+## Apple `bookmarkData` format
+
+Apple's `bookmarkData` format is used to encode information that can be resolved
+into a `URL` object for a file even if the user moves or renames it. Can also
+contain security scoping information for App Sandbox support.
+
+These `bookmarkData` blobs are often found endcoded in data fields of Binary
+Property Lists. Notable examples include:
+
+- `com.apple.finder.plist` - contains an `FXRecentFolders` value, which is an
+  array of ten objects, each of which consists of a `name` and `file-bookmark`
+  field, which is a `bookmarkData` object for each recently accessed folder
+  location.
+
+- `com.apple.LSSharedFileList.RecentApplications.sfl2` - `sfl2` files are
+  actually `plist` files of the `NSKeyedArchiver` format. They can be parsed the
+  same as `plist` files, but they have a more complicated tree-like structure
+  than would typically be found, which can make locating and retrieving specific
+  values difficult, even once it has been converted to a JSON representation.
+  For more information about these types of files, see 
+  [Sarah Edwards'](https://www.mac4n6.com/blog/2016/1/1/manual-analysis-of-nskeyedarchiver-formatted-plist-files-a-review-of-the-new-os-x-1011-recent-items)
+  excellent research on the subject.
+
+Locating `bookmarkData` objects in plist files, especially those of the
+`NSKeyedArchiver` type, is a place where `fq`'s recursive searching really
+shines. Here, we recursively descend through the decoded tree, probing for and 
+selecting any `bookmark` blobs, then converting them to readable JSON with
+`torepr`:
+```
+fq '.. | select(format=="bookmark") | .map(. | torepr)' com.apple.LSSharedFileList.RecentApplications.sfl2
```

Comment:
I've noticed the exact same thing and I'm not sure what to make of it yet. I'll do some digging and see if I can understand the problem.

Reply:
> This example looks a bit strange, is it suppose to work with the current code? looks like it decodes `com.apple.LSSharedFileList.RecentApplications.sfl2` as bplist and recurses looking for things decoded as `bookmark`, is the plan to make bplist decoder somehow detect and decode booktmark?

Yeah, was looking for a quick one-liner for recovering bookmarks from any given `plist`/`sfl2` file, since that would be a big use case for forensic analysts.

Reply:
> Yeah, was looking for a quick one-liner for recovering bookmarks from any given `plist`/`sfl2` file, since that would be a big use case for forensic analysts.

I see, make sense, should be possible. Would like the example to output one big JSON with all bookmark blob expanded "in place" in the sfl2 tree or output each bookmark blob as a separate JSON fragment? or something else?

Reply:
It would be good to find both solutions. I had considered maybe writing some shell functions/aliases and putting them in a separate repo as a kind of `fq` forensics toolkit. I like the first option you presented, displaying them expanded in place in the `sfl2` tree. As for how to display just the bookmark data, I'm not sure if it makes more sense to display them as separate JSON fragments, or as an array of objects.

Reply:
> It would be good to find both solutions. I had considered maybe writing some shell functions/aliases and putting them in a separate repo as a kind of `fq` forensics toolkit. I like the first option you presented, displaying them expanded in place in the `sfl2` tree. As for how to display just the bookmark data, I'm not sure if it makes more sense to display them as separate JSON fragments, or as an array of objects.

fq forensics toolkit sounds great! maybe a good place to test things and if stable/reusable/generic enough some things can move into fq? i've actually thought about doing something similar but for media files, i have a bunch of mp4 related snippets that don't really fit into fq.

I got a bit excited about decoding the NSKeysArchived and made some progress. Still very rough:

This is with the snake_case fix for apple_bookmark torepr.

Put this in ns_keyed_archiver.jq
```
def from_ns_keyed_archiver:
  (  . as {"$objects": $objs, "$top": {root: $root_uid}}
  | def _f($id):
      ( . #debug({$id})
      | $objs[$id]
      # | debug
      | if type == "string" then .
        elif type == "number" then .
        else
          (. as {"$class": $class}
          | if $class == 13 then # NSDictionary?
              ( . as {"NS.keys": $ns_keys, "NS.objects": $ns_objects}
              | [$ns_keys, $ns_objects]
              | transpose
              | map(
                  ( . as [$k, $o]
                  | {key: _f($k), value: _f($o)}
                  )
                )
              # | debug
              | from_entries
              )
            elif $class == 58 then #?
              ( . as {"NS.objects": $ns_objects}
              | $ns_objects
              | map(_f(.))
              )
            else "class-\($class)"
            end
          )
        end
      );
    _f($root_uid)
  );
```

Now this works, lots of options and things but nice to see it seems to work. Does it look correct?
```
$ go run . -o bits_format=string -L . 'include "ns_keyed_archiver"; torepr | from_ns_keyed_archiver | (.. | select(.Bookmark)?.Bookmark) |= (apple_bookmark | torepr)' '/Users/wader//Library/Application Support/com.apple.sharedfilelist/com.apple.LSSharedFileList.RecentApplications.sfl2'
{
  "items": [
    {
      "Bookmark": {
        "983055": true,
        "creation_options": 536870912,
        "display_name": "Firefox.app",
        "target_cnid_path": [
          77548117,
          1260261
        ],
        "target_creation_date": 532186965,
        "target_flags": "\u0012\u0002\u0000\u0000\u0000\u0000\u0000\u0000\u001f\u0002\u0000\u0000\u0000\u0000\u0000\u0000\u001a\u0002\u0000\u0000\u0000\u0000\u0000\u0000",
        "target_path": [
          "Applications",
          "Firefox.app"
        ],
        "volume_creation_date": 687789381,
        "volume_flags": "\ufffd\u0000\u0000\u0000\u0001\u0000\u0000\u0000\ufffd\u0013\u0000\u0000\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000",
        "volume_is_root": true,
        "volume_name": "Macintosh HD",
        "volume_path": "/",
        "volume_size": 250685575168,
        "volume_url": "file:///",
        "volume_uuid": "D63F4409-BCCD-48AE-896E-317E982AD80E"
      },
      "CustomItemProperties": {},
      "Name": "Firefox.app",
      "uuid": "8FD3DE25-8DCD-4B3B-915C-1CD6BE90AFED",
      "visibility": 0
    },
... more items here ...
}
```

Reply:
This is amazing, going to make some people very happy 😄

Reply:
Only thing is `elif $class == 58 then #?` should be `elif $class == 56 then #?`, unless they have changed the format. Had to tweak it to make it work on my machine.

Reply:
Seems like those class numbers might cause problems, unfortunately there is some variability. I've had to make tweaks to make it work on different files.

Reply:
Seems like those class numbers might cause problems, unfortunately there is some variability. This solution seems to be more flexible: 
```
def from_ns_keyed_archiver:
  (  . as {"$objects": $objs, "$top": {root: $root_uid}}
  | def _f($id):
      ( . #| debug({$id})
      | $objs[$id]
      #| debug
      | if type == "string" then .
        elif type == "number" then .
        else
          (. as {"$class": $class}
          | debug
          | if ."NS.keys" != null and ."NS.objects" != null then
              ( . as {"NS.keys": $ns_keys, "NS.objects": $ns_objects}
              | [$ns_keys, $ns_objects]
              | transpose
              | map
                (
                  ( . as [$k, $o]
                  | {key: _f($k), value: _f($o)}
                  )
                )
              | from_entries
              )
            elif ."NS.objects" != null then
              ( . as {"NS.objects": $ns_objects}
              | $ns_objects
              | map(_f(.))
              )
            else "class-\($class)"
            end
          )
        end
      );
    _f($root_uid)
  );
```

Reply:
```
fq -o bits_format=string -L . 'include "ns_keyed_archiver"; torepr | from_ns_keyed_archiver | (.. | select(.Bookmark)?.Bookmark) |= (apple_bookmark | torepr)'
```
Some variation of this would be great as an alias, maybe `sfl2_bookmarks`? I'm open to ideas about how you want to structure snippets. It might make sense to put them into a separate repository so that people that want to use them but are running `fq` via brew/go install/etc can clone the repo without bringing it all of the other stuff.

Reply:
Do we want to put `ns_keyed_archiver.jq` in the `bplist` package? And if so, do we need to register the function in the same way that we do `torepr`?

Reply:
I think we can include ns_keyed_archiver.jq for bplist, it seem to be very common usage for plists? but it feels a bit rough atm, maybe we can work on it separate to this PR? for example i wonder how these class numbers work, are they stable? maybe the ones from cocoa at least? if not i guess we have to have an argument etc that is used to select mapping?

There is no need to register functions, the will just appear. But `torepr` is a bit special as it is overloaded per format, if your curious it's implemented here https://github.com/wader/fq/blob/012e7527f05508e1d154a9cf18de4e9b8b216686/pkg/interp/funcs.jq#L159 with help of https://github.com/wader/fq/blob/master/pkg/interp/format_func.jq

Reply:
Ok, I'll set it aside for the moment and open an issue for it.